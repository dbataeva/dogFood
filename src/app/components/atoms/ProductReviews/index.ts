export { ProductReviews } from './ProductReviews.component';
