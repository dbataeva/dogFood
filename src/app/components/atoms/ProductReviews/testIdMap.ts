export const TEST_ID_MAP = {
	reviewsList: 'reviewsList',
	review: 'review',
	deleteReviewButton: 'deleteReviewButton',
};
