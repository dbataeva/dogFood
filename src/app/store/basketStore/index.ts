export {
	basketReducer,
	addProductToBasket,
	deleteProductFromBasket,
	sliceName as basketSliceName,
	decreaseNumberOfProductsInBasket,
} from './basketStore';
