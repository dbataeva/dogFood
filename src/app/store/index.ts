export {
	selectUser,
	selectBasket,
	useAppDispatch,
	useAppSelector,
} from './appStore';

export { setUserData } from './userStore';

export {
	addProductToBasket,
	deleteProductFromBasket,
	decreaseNumberOfProductsInBasket,
} from './basketStore';
