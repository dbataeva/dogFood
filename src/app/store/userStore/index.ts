export {
	userReducer,
	setUserData,
	sliceName as userSliceName,
} from './userStore';
