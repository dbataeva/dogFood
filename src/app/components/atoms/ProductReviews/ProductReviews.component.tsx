import { FC, Fragment, memo, useEffect, useCallback } from 'react';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import Divider from '@mui/material/Divider';
import ListItemText from '@mui/material/ListItemText';
import ListItemAvatar from '@mui/material/ListItemAvatar';
import Avatar from '@mui/material/Avatar';
import { Typography, IconButton } from '@mui/material';
import { DeleteForever } from '@mui/icons-material';

import { Product } from '../../../../types';
import {
	fetchUserInfo,
	selectUser,
	useAppDispatch,
	useAppSelector,
} from '../../../store';
import { useDeleteReviewMutation } from '../../../../api';
import { TEST_ID_MAP } from './testIdMap';

type ProductReviewsProps = {
	reviews?: Product['reviews'];
	productId: string;
};

export const ProductReviews: FC<ProductReviewsProps> = memo(
	({ productId, reviews }) => {
		const dispatch = useAppDispatch();
		const { currentUser } = useAppSelector(selectUser);
		const [deleteReviewFn] = useDeleteReviewMutation();

		const clickDeleteReviewHandler = useCallback(
			(id: string) => {
				return () => deleteReviewFn({ productId, reviewId: id });
			},
			[deleteReviewFn, productId]
		);

		useEffect(() => {
			!currentUser && dispatch(fetchUserInfo());
		}, []);

		return (
			<List
				data-testid={TEST_ID_MAP.reviewsList}
				sx={{
					width: '100%',
					maxWidth: '992px',
					marginInline: 'auto',
					bgcolor: 'background.paper',
				}}>
				{reviews?.map((review) => {
					const deleteReview = clickDeleteReviewHandler(review._id);

					return (
						<Fragment key={review._id}>
							<ListItem
								alignItems='flex-start'
								data-testid={TEST_ID_MAP.review}>
								<ListItemAvatar>
									<Avatar alt='Remy Sharp' src={review.author.avatar} />
								</ListItemAvatar>
								<ListItemText
									primary={review.author.name}
									secondary={
										<>
											<Typography
												sx={{ display: 'inline' }}
												component='span'
												variant='body2'
												color='text.primary'>
												{review.text}
											</Typography>
										</>
									}
								/>
								{review.author._id === currentUser?._id && (
									<IconButton
										onClick={deleteReview}
										data-testid={TEST_ID_MAP.deleteReviewButton}>
										<DeleteForever />
									</IconButton>
								)}
							</ListItem>
							<Divider variant='inset' component='li' />
						</Fragment>
					);
				})}
			</List>
		);
	}
);

ProductReviews.displayName = 'ProductReviews';
