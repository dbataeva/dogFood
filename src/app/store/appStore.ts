import { configureStore } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import type { TypedUseSelectorHook } from 'react-redux';

import { api, authApi, productsApi, userApi } from '../../api';
import { userSliceName, userReducer } from './userStore';
import { basketReducer, basketSliceName } from './basketStore';

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export const store = configureStore({
	reducer: {
		user: userReducer,
		basket: basketReducer,
		[authApi.reducerPath]: authApi.reducer,
		[userApi.reducerPath]: userApi.reducer,
		[productsApi.reducerPath]: productsApi.reducer,
	},
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware({
			thunk: {
				extraArgument: api,
			},
		}).concat(authApi.middleware, productsApi.middleware, userApi.middleware),
	devTools: true,
});

export const selectUser = (state: RootState) => state[userSliceName];
export const selectBasket = (state: RootState) => state[basketSliceName];
