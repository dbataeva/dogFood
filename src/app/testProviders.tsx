import { RouterProvider, createMemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { routerConfig } from './constants';
import { store } from './store/appStore';
import * as AuthorizedUtils from './utils/getIsUserAuthorized';
import { FC, ReactNode } from 'react';

export const testProvidersFn = (initialPath = '/', isAuthorized?: boolean) => {
	const router = createMemoryRouter(routerConfig, {
		initialEntries: [initialPath],
	});

	jest
		.spyOn(AuthorizedUtils, 'getIsUserAuthorized')
		.mockReturnValue(isAuthorized ? true : false);

	render(
		<Provider store={store}>
			<RouterProvider router={router} />
		</Provider>
	);

	return router;
};

export const TestProvidersFC: FC<{ children: ReactNode }> = ({ children }) => {
	return <Provider store={store}>{children}</Provider>;
};
