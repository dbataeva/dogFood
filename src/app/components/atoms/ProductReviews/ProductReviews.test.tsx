import { screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { userEvent } from '@testing-library/user-event';
import ReactTestRenderer from 'react-test-renderer';

import * as productsApi from '../../../../api/productsApi';
import { TEST_ID_MAP } from './testIdMap';
import { TestProvidersFC, testProvidersFn } from '../../../testProviders';
import * as appStore from '../../../store/appStore';
import { ProductReviews } from './ProductReviews.component';

const mockedProductWithoutReview = {
	discount: 0,
	stock: 10,
	available: true,
	pictures: 'https://react-learning.ru/image-compressed/12.jpg',
	likes: ['64107e09aa397121838f2926'],
	reviews: [],
	tags: [],
	isPublished: true,
	_id: '456',
	name: 'Печенья с яблоком (печень говяжья, сердце)',
	author: {
		name: 'Максим',
		about: 'Ментор',
		avatar:
			'https://nankeluxuryhomesprescott.com/wp-content/uploads/2019/03/PlaceholderImageMale.jpg',
		_id: '622bd81b06c7d323b8ae4614',
		email: 'someEmail@inbox.ru',
		__v: 0,
	},
	price: 340,
	wight: '14 шт',
	description: 'Описание demo',
	created_at: '2022-03-12T10:38:04.499Z',
	updated_at: '2023-11-19T13:16:33.364Z',
	__v: 0,
};

const mockedProductWithCurrentUserReview = {
	discount: 0,
	stock: 10,
	available: true,
	pictures: 'https://react-learning.ru/image-compressed/12.jpg',
	likes: ['64107e09aa397121838f2926'],
	reviews: [
		{
			rating: 5,
			_id: '645182088fbc473fa8a23e42',
			text: 'Понравилось',
			author: {
				name: 'Александр Филимонов',
				about: 'Frontend10-student',
				avatar:
					'https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80',
				_id: '123',
				email: 'somaEmail.av@ya.ru',
				group: 'group-10',
				__v: 0,
			},
			product: '622c780c77d63f6e70967d27',
			created_at: '2023-05-02T21:35:04.343Z',
			updated_at: '2023-05-02T21:35:04.343Z',
			__v: 0,
		},
	],
	tags: [],
	isPublished: true,
	_id: '456',
	name: 'Печенья с яблоком (печень говяжья, сердце)',
	author: {
		name: 'Максим',
		about: 'Ментор',
		avatar:
			'https://nankeluxuryhomesprescott.com/wp-content/uploads/2019/03/PlaceholderImageMale.jpg',
		_id: '622bd81b06c7d323b8ae4614',
		email: 'someEmail@inbox.ru',
		__v: 0,
	},
	price: 340,
	wight: '14 шт',
	description: 'Описание demo',
	created_at: '2022-03-12T10:38:04.499Z',
	updated_at: '2023-11-19T13:16:33.364Z',
	__v: 0,
};

const mockedProductWithAnotherUserReview = {
	discount: 0,
	stock: 10,
	available: true,
	pictures: 'https://react-learning.ru/image-compressed/12.jpg',
	likes: ['64107e09aa397121838f2926'],
	reviews: [
		{
			rating: 3,
			_id: '645182088fbc473fa8a23e41',
			text: 'Не понравилось',
			author: {
				name: 'Александра Филимонова',
				about: 'Frontend10-student',
				avatar:
					'https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80',
				_id: '63fdee524ee419975fbd29e3',
				email: 'someEmail.av@ya.ru',
				group: 'group-10',
				__v: 0,
			},
			product: '622c780c77d63f6e70967d27',
			created_at: '2023-05-02T21:35:04.343Z',
			updated_at: '2023-05-02T21:35:04.343Z',
			__v: 0,
		},
	],
	tags: [],
	isPublished: true,
	_id: '456',
	name: 'Печенья с яблоком (печень говяжья, сердце)',
	author: {
		name: 'Максим',
		about: 'Ментор',
		avatar:
			'https://nankeluxuryhomesprescott.com/wp-content/uploads/2019/03/PlaceholderImageMale.jpg',
		_id: '622bd81b06c7d323b8ae4614',
		email: 'someEmail@inbox.ru',
		__v: 0,
	},
	price: 340,
	wight: '14 шт',
	description: 'Описание demo',
	created_at: '2022-03-12T10:38:04.499Z',
	updated_at: '2023-11-19T13:16:33.364Z',
	__v: 0,
};

const productIdWIthNumberOfReview = '456';

const mockedProductWithNumberOfReview = {
	discount: 0,
	stock: 10,
	available: true,
	pictures: 'https://react-learning.ru/image-compressed/12.jpg',
	likes: ['64107e09aa397121838f2926'],
	reviews: [
		{
			rating: 3,
			_id: '645182088fbc473fa8a23e41',
			text: 'Не понравилось',
			author: {
				name: 'Александра Филимонова',
				about: 'Frontend10-student',
				avatar:
					'https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80',
				_id: '63fdee524ee419975fbd29e3',
				email: 'someEmail.av@ya.ru',
				group: 'group-10',
				__v: 0,
			},
			product: '622c780c77d63f6e70967d27',
			created_at: '2023-05-02T21:35:04.343Z',
			updated_at: '2023-05-02T21:35:04.343Z',
			__v: 0,
		},
		{
			rating: 3,
			_id: '645182088fbc473fa8a23e41',
			text: 'Не понравилось',
			author: {
				name: 'Александра Филимонова',
				about: 'Frontend10-student',
				avatar:
					'https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80',
				_id: '63fdee524ee419975fbd29e3',
				email: 'someEmail.av@ya.ru',
				group: 'group-10',
				__v: 0,
			},
			product: '622c780c77d63f6e70967d27',
			created_at: '2023-05-02T21:35:04.343Z',
			updated_at: '2023-05-02T21:35:04.343Z',
			__v: 0,
		},
	],
	tags: [],
	isPublished: true,
	_id: productIdWIthNumberOfReview,
	name: 'Печенья с яблоком (печень говяжья, сердце)',
	author: {
		name: 'Максим',
		about: 'Ментор',
		avatar:
			'https://nankeluxuryhomesprescott.com/wp-content/uploads/2019/03/PlaceholderImageMale.jpg',
		_id: '622bd81b06c7d323b8ae4614',
		email: 'someEmail@inbox.ru',
		__v: 0,
	},
	price: 340,
	wight: '14 шт',
	description: 'Описание demo',
	created_at: '2022-03-12T10:38:04.499Z',
	updated_at: '2023-11-19T13:16:33.364Z',
	__v: 0,
};

describe('Отзывы о продукте', () => {
	beforeEach(() => {
		jest.spyOn(appStore, 'selectUser').mockReturnValue({
			isLoading: false,
			error: null,
			currentUser: {
				name: 'Александр Филимонов',
				about: 'Frontend10-student',
				avatar:
					'https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80',
				_id: '123',
				email: 'somaEmail.av@ya.ru',
				group: 'group-10',
				__v: 0,
			},
		});
	});

	it('должны рендериться', () => {
		jest.spyOn(productsApi, 'useGetProductByIdQuery').mockReturnValue({
			isLoading: false,
			error: undefined,
			isError: false,
			data: mockedProductWithoutReview,
			refetch: jest.fn(),
		});

		testProvidersFn('/products/123', true);

		expect(screen.getByTestId(TEST_ID_MAP.reviewsList)).toBeInTheDocument();
	});

	it('должно рендериться соответствующее кол-во отзывов', () => {
		jest.spyOn(productsApi, 'useGetProductByIdQuery').mockReturnValue({
			isLoading: false,
			error: undefined,
			isError: false,
			data: mockedProductWithNumberOfReview,
			refetch: jest.fn(),
		});

		testProvidersFn('/products/123', true);

		expect(screen.getAllByTestId(TEST_ID_MAP.review).length).toEqual(
			mockedProductWithNumberOfReview.reviews.length
		);
	});

	describe('удаление отзыва', () => {
		it('должно быть доступно, если отзыв принадлежит текущему пользователю', () => {
			jest.spyOn(productsApi, 'useGetProductByIdQuery').mockReturnValue({
				isLoading: false,
				error: undefined,
				isError: false,
				data: mockedProductWithCurrentUserReview,
				refetch: jest.fn(),
			});

			testProvidersFn('/products/123', true);

			expect(
				screen.getByTestId(TEST_ID_MAP.deleteReviewButton)
			).toBeInTheDocument();
		});

		it('должно быть недоступно, если отзыв принадлежит другому пользователю', () => {
			jest.spyOn(productsApi, 'useGetProductByIdQuery').mockReturnValue({
				isLoading: false,
				error: undefined,
				isError: false,
				data: mockedProductWithAnotherUserReview,
				refetch: jest.fn(),
			});

			testProvidersFn('/products/123', true);

			expect(
				screen.queryByTestId(TEST_ID_MAP.deleteReviewButton)
			).not.toBeInTheDocument();
		});

		it('должен удаляться отзыв при клике на кнопку удаления', () => {
			const deleteReview = jest.fn();

			jest
				.spyOn(productsApi, 'useDeleteReviewMutation')
				.mockReturnValue([deleteReview, {} as any]);

			jest.spyOn(productsApi, 'useGetProductByIdQuery').mockReturnValue({
				isLoading: false,
				error: undefined,
				isError: false,
				data: mockedProductWithCurrentUserReview,
				refetch: jest.fn(),
			});

			testProvidersFn('/products/123', true);

			const deleteReviewButton = screen.getByTestId(
				TEST_ID_MAP.deleteReviewButton
			);

			userEvent.click(deleteReviewButton);

			waitFor(() => expect(deleteReview).toHaveBeenCalled());
		});
	});

	it("должен соответствовать snapshot'у", () => {
		const tree = ReactTestRenderer.create(
			<TestProvidersFC>
				<ProductReviews
					productId={productIdWIthNumberOfReview}
					reviews={mockedProductWithNumberOfReview.reviews}
				/>
			</TestProvidersFC>
		).toJSON();

		expect(tree).toMatchSnapshot();
	});
});
