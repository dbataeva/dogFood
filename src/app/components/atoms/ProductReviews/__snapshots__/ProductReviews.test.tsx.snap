// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Отзывы о продукте должен соответствовать snapshot'у 1`] = `
<ul
  className="MuiList-root MuiList-padding css-gcof6a-MuiList-root"
  data-testid="reviewsList"
>
  <li
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
    data-testid="review"
    disabled={false}
  >
    <div
      className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
      >
        <img
          alt="Remy Sharp"
          className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80"
        />
      </div>
    </div>
    <div
      className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        Александра Филимонова
      </span>
      <p
        className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
        >
          Не понравилось
        </span>
      </p>
    </div>
  </li>
  <li
    className="MuiDivider-root MuiDivider-inset css-uls1yk-MuiDivider-root"
    role="separator"
  />
  <li
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
    data-testid="review"
    disabled={false}
  >
    <div
      className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
      >
        <img
          alt="Remy Sharp"
          className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://images.unsplash.com/photo-1552944150-6dd1180e5999?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=725&q=80"
        />
      </div>
    </div>
    <div
      className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        Александра Филимонова
      </span>
      <p
        className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
        >
          Не понравилось
        </span>
      </p>
    </div>
  </li>
  <li
    className="MuiDivider-root MuiDivider-inset css-uls1yk-MuiDivider-root"
    role="separator"
  />
</ul>
`;
